### PR TITLE
store: do not resume a download when we already have the whole thing

### PR DESCRIPTION
**Pull request by chipaca**
Without this a .partial file that has been completely downloaded still
triggers a request to the server (which will then come back with a
416).

This lets you drop a .snap in /var/lib/snapd/snaps as a .partial,
which could be used to speed up tests. It'll still be checksummed
(twice), but it won't hit the network for the download unless the
store says it's got a delta.

Doing the same thing for delta downloads is left as an exercise to the
reader.

### CODE REVIEW

**Comment by chipaca** [comment]
@zyga better? :-)

**Comment by zyga** [comment]
Much better, thank you!

File: store/store.go [hunk 1]
```diff
@@ -1359,7 +1359,22 @@ func (s *Store) Download(ctx context.Context, name string, targetPath string, do
 		url = downloadInfo.DownloadURL
 	}
 
-	err = download(ctx, name, downloadInfo.Sha3_384, url, user, s, w, resume, pbar)
+	if downloadInfo.Size == 0 || resume < downloadInfo.Size {
+		err = download(ctx, name, downloadInfo.Sha3_384, url, user, s, w, resume, pbar)
+	} else if downloadInfo.Sha3_384 != "" {
```
Comment (pedronis) [inline store/store.go]: this is weird, shouldn't this always be the case ? and if not, shouldn't we error earlier?

File: store/store.go [hunk 2]
```diff
@@ -1359,7 +1359,22 @@ func (s *Store) Download(ctx context.Context, name string, targetPath string, do
 		url = downloadInfo.DownloadURL
 	}
 
-	err = download(ctx, name, downloadInfo.Sha3_384, url, user, s, w, resume, pbar)
+	if downloadInfo.Size == 0 || resume < downloadInfo.Size {
```
Comment (pedronis) [inline store/store.go]: are we sure resume cannot be > ?

Comment (chipaca) [inline store/store.go]: I know it can, but it should then fail the checksum

File: store/store.go (hunk 1 above)
Comment (chipaca) [inline store/store.go]: there are tests that don't provide the checksum

Comment (pedronis) [inline store/store.go]: yes, maybe, but you changed real runtime code, the old code was always downloading, something seems off now

Comment (pedronis) [inline store/store.go]: what I mean, if you think we cannot error, it seems more convervative to behave as before in the case == "", not do a random thing as with this code

Comment (chipaca) [inline store/store.go]: gah, turns out the only test that didn't set it that mattered was my own. fixed.

**Pull request merged**